### ISSUE TITLE
Error: Container is running beyond physical memory limits

### ISSUE THREAD

**Issue by clifff** [awslabs/athena-glue-service-logs#19]
I'm trying to run some analysis on a collection of S3 Access Logs, and set up a Glue job using the steps in the README to do so. The set of logs is about 14 GB over  12.8 million files.  Whenever I kick off the job, it runs for about 13 minutes and then fails with a `Command failed with exit code 1` message. Looking at the logs, I see this line that seems important:

```
Diagnostics: Container [pid=11027,containerID=container_1569865532923_0001_01_000001] is running beyond physical memory limits. Current usage: 5.5 GB of 5.5 GB physical memory used; 7.7 GB of 27.5 GB virtual memory used. Killing container.
```

This is corroborated by CloudWatch metrics, which show the driver memory usage steadily climbing and the executor staying low.

[Based on the `athena_glue_service_logs` blog post here](https://aws.amazon.com/blogs/big-data/easily-query-aws-service-logs-using-amazon-athena/), it seems like my volume of data is well within the expected limits. I retried the job after adding the `--conf` parameter set to `spark.yarn.executor.memoryOverhead=1G`, but it failed in the same way.

Any advice for getting this to work are appreciated - otherwise I'll follow the [Glue documentation suggestion](https://docs.aws.amazon.com/glue/latest/dg/monitor-profile-debug-oom-abnormalities.html) of writing a script to do the conversion using DynamicFrames.

**Comment by dacort** [comment]
Hi @clifff - there's a couple things you can try here.

1) Change the worker type in the Glue job to one with larger memory (G.1X - see screenshot below). 
2) Try increasing the `spark.yarn.executor.memoryOverhead` even more, but there is only so far you can go with that. 
3) You can also try increasing the driver memory, since you mention that's increasing. Set a parameter with the key `--conf` and value `spark.driver.memory=10g`.

I would recommend trying the first option as that will inherently give you more memory to work with.

![image](https://user-images.githubusercontent.com/1512/65909645-ea219f80-e37d-11e9-9580-b788a4622935.png)

**Comment by clifff** [comment]
Thanks for the tip @dacort! Didn't realize worker type was configurable like that. I upped to `G.1X` and let the job run again - churned for about 100 minutes before crashing again. Found this in the logs:

```
Log Contents:
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 169.254.169.254
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 169.254.169.254
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): glue.us-east-1.amazonaws.com
INFO:athena_glue_service_logs.job:Recurring run, only looking for recent partitions on raw catalog.
#
# java.lang.OutOfMemoryError: Java heap space
# -XX:OnOutOfMemoryError="kill -9 %p"
# Executing /bin/sh -c "kill -9 10754"...
os::fork_and_exec failed: Cannot allocate memory (12)
End of LogType:stdout
```

Which matches what Cloudwatch is showing:

<img width="1220" alt="Screen Shot 2019-09-30 at 4 32 31 PM" src="https://user-images.githubusercontent.com/51312/65918849-58775980-e3a0-11e9-8608-5f1df1e463e6.png">

It seems promising it didn't hit a memory usage of `1` and immediately crash, but that does make me think it's not necessary to configure the driver to a specific amount. Went ahead and  raised `spark.yarn.executor.memoryOverhead` to `2G` and will try that out.

**Comment by clifff** [comment]
Confirm, timed out in about the same amount of time with the `2G` setting.

**Comment by dacort** [comment]
OK, thanks for trying that @clifff - looks like building up the list of those 13M files is taking up quite the resources. Give me a few days to see if I can reproduce this in my own environment to see what options there might be. There's definitely still some more testing for these scripts at that scale.

**Comment by clifff** [comment]
Sounds good - thanks for looking into this @dacort! Happy to tweak settings/code and retry whenever.

**Comment by dacort** [comment]
Hey @clifff - Unfortunately haven't been able to take a look much deeper. How high did you bump `spark.driver.memory`?

A couple other options:
- Try increasing executor memory as well (don't think this is the issue tho): `--conf` `spark.yarn.executor.memory=1g` and keep increasing
- Take a look at this [doc about grouping input files](https://docs.aws.amazon.com/glue/latest/dg/grouping-input-files.html). You'll have to modify `converter.py` and I'm not sure what the options are to read from the catalog, but you can test reading the options with an explicit S3 path to at least see if it works.

There is some more detail on debugging OOM issues here as well: https://docs.aws.amazon.com/glue/latest/dg/monitor-profile-debug-oom-abnormalities.html#monitor-profile-debug-oom-driver

**Comment by clifff** [comment]
No worries! I actually was successful loading the logs onto an EC2 instance. Turns out the bucket inventory size was way off and it was more like 60 GB of logs... but the good news is I was able to filter it down to ~100 mb of relevant lines using `ripgrep`, and got the info I needed from there.

Will go ahed and close this for now since, but feel free to re-open if you want to track the issue further.

**Comment by dacort** [comment]
👍 Sounds good, thanks!

**Comment by dacort** [comment]
I didn't realize you were just trying to do a one-time query. For future reference, this library creates two tables - one for the "raw" unconverted data and another for the "optimized" parquet data. This appears to have been failing during the conversion process, but you still could have queried the raw data. But `ripgrep` for the win! One of my favorites tools.

**Comment by RickardCardell** [comment]
Hi
So I've got a similar issue, that I couldn't make the job run on not that large dataset: 25GB s3 access logs/day for 30days. 

I've tried with 50 G.2X worker nodes and high memory settings to no avail.

I instead went to the code and skipped the repartition stage: https://github.com/awslabs/athena-glue-service-logs/blob/master/athena_glue_service_logs/converter.py#L66

The job succeeded with 100 'standard' workers after only 4hours.
The drawback is of course that more objects were created: between 50-140 per day-partition.

Would it make sense to make the repartitioning configurable?
